alternator/ttl.cc

```diff
@@ -730,34 +729,88 @@ static future<bool> scan_table(
     expiration_stats.scan_table++;
     // FIXME: need to pace the scan, not do it all at once.
     scan_ranges_context scan_ctx{s, proxy, std::move(column_name), std::move(member)};
-    auto erm = db.real_database().find_keyspace(s->ks_name()).get_vnode_effective_replication_map();
-    auto my_host_id = erm->get_topology().my_host_id();
-    token_ranges_owned_by_this_shard my_ranges(s, co_await ranges_holder_primary::make(erm, my_host_id));
-    while (std::optional<dht::partition_range> range = my_ranges.next_partition_range()) {
-        // Note that because of issue #9167 we need to run a separate
-        // query on each partition range, and can't pass several of
-        // them into one partition_range_vector.
-        dht::partition_range_vector partition_ranges;
-        partition_ranges.push_back(std::move(*range));
-        // FIXME: if scanning a single range fails, including network errors,
-        // we fail the entire scan (and rescan from the beginning). Need to
-        // reconsider this. Saving the scan position might be a good enough
-        // solution for this problem.
-        co_await scan_table_ranges(proxy, scan_ctx, std::move(partition_ranges), abort_source, page_sem, expiration_stats);
-    }
-    // If each node only scans its own primary ranges, then when any node is
-    // down part of the token range will not get scanned. This can be viewed
-    // as acceptable (when the comes back online, it will resume its scan),
-    // but as noted in issue #9787, we can allow more prompt expiration
-    // by tasking another node to take over scanning of the dead node's primary
-    // ranges. What we do here is that this node will also check expiration
-    // on its *secondary* ranges - but only those whose primary owner is down.
-    token_ranges_owned_by_this_shard my_secondary_ranges(s, co_await ranges_holder_secondary::make(erm, my_host_id, gossiper));
-    while (std::optional<dht::partition_range> range = my_secondary_ranges.next_partition_range()) {
-        expiration_stats.secondary_ranges_scanned++;
-        dht::partition_range_vector partition_ranges;
-        partition_ranges.push_back(std::move(*range));
-        co_await scan_table_ranges(proxy, scan_ctx, std::move(partition_ranges), abort_source, page_sem, expiration_stats);
+    auto is_replica_alive_fn = [&gossiper] (locator::host_id host_id) -> bool { return gossiper.is_alive(host_id); };
```

Comment:
Nitpick: if I understand correctly, is_replica_alive_fn is only used in the uses_tablets() case, so I think it's better to move it inside the following if (and even better, move it even closer to where it's actually used).